hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/JDBCSource.java

```diff
@@ -0,0 +1,233 @@
+              dataFrameReader = dataFrameReader.option(kv[0], kv[1]);
+              LOG.info("{} = {} has been set for JDBC pull ", kv[0], kv[1]);
+            }
+          }
+        }
+      }
+      if (properties.getBoolean(Config.IS_INCREMENTAL)) {
+        DataSourceUtils.checkRequiredProperties(properties, Arrays.asList(Config.INCREMENTAL_COLUMN));
+      }
+      return dataFrameReader;
+    } catch (Exception e) {
+      throw new HoodieException(e);
+    } finally {
+      IOUtils.closeStream(passwordFileStream);
+    }
+  }
+
+  @Override
+  protected Pair<Option<Dataset<Row>>, String> fetchNextBatch(Option<String> lastCkptStr, long sourceLimit) {
+    try {
+      DataSourceUtils.checkRequiredProperties(props, Arrays.asList(Config.URL, Config.DRIVER_CLASS, Config.USER,
+          Config.RDBMS_TABLE_NAME, Config.IS_INCREMENTAL));
+
+      Option<String> beginInstant =
+          lastCkptStr.isPresent() ? lastCkptStr.get().isEmpty() ? Option.empty() : lastCkptStr : Option.empty();
+      boolean isIncremental = props.getBoolean(Config.IS_INCREMENTAL);
+      if (beginInstant.equals(Option.empty())) {
+        LOG.info("No previous checkpoints found.. ");
+        Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props).load();
+        return sendDfAndCheckpoint(rowDataset, isIncremental);
+      } else {
+        if (StringUtils.isNullOrEmpty(beginInstant.get())) {
+          LOG.warn("Previous checkpoint entry was null or empty. Falling back to full jdbc pull.");
+          Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props).load();
+          return sendDfAndCheckpoint(rowDataset, isIncremental);
+        } else {
+          String query = String
+              .format(ppdQuery, props.getString(Config.RDBMS_TABLE_NAME), props.getString(Config.INCREMENTAL_COLUMN),
+                  beginInstant.get());
+          LOG.info("Referenced last checkpoint and prepared new predicate pushdown query for jdbc pull {}", query);
+          Dataset<Row> rowDataset = validatePropsAndGetDataFrameReader(sparkSession, props)
+              .option(props.getString(Config.RDBMS_TABLE_PROP), query).load();
+          return sendDfAndCheckpoint(rowDataset, isIncremental);
+
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Exception while running JDBCSource ", e);
+      return Pair.of(Option.empty(), null);
+    }
+  }
+
+  @NotNull
+  private Pair<Option<Dataset<Row>>, String> sendDfAndCheckpoint(Dataset<Row> rowDataset, boolean isIncremental) {
+    if (isIncremental) {
+      Column incrementalColumn = rowDataset.col(props.getString(Config.INCREMENTAL_COLUMN));
+      final String max = rowDataset.agg(functions.max(incrementalColumn).cast(DataTypes.StringType)).first()
+          .getString(0);
+      LOG.info("Sending {} with checkpoint val {} ", incrementalColumn, max);
+      return Pair.of(Option.of(rowDataset), max);
+    } else {
```

Comment:
okay. so the use of `isIncremental` is to actually to reset the checkpoints once in a while and force a bulk load?

Reply:
Let me put it this way for you. 
case 1: A JDBC job can be continuous but not incremental
case 2: A JDBC job can be continuous and incremental
case 3: A JDBC job is not continuous but is incremental
case 4: A JDBC job is not continuous and not incremental

In case 1 and case 4 we never reference anything to/from the checkpoint and we pull the entire table always.

In case 2 and 3 we reference the checkpoint to see the last checkpointed value. If the value is present, we build an incremental query or else we do the whole table pull and write the max of incremental column and keep doing this same thing again and again. This is what I think will be best.

Reply:
makes sense.. Once you have this, we should also probably think about generalizing across sources.. I like the fact that you can do both modes in the same way. 

Please file a JIRA for tracking this, if you also agree